tempesta_fw/sock.c

```diff
@@ -1036,45 +1039,283 @@ ss_connect(struct sock *sk, struct sockaddr *uaddr, int uaddr_len, int flags)
 		goto out;
 	err = 0;
 out:
-	release_sock(sk);
+	bh_unlock_sock(sk);
 	return err;
 }
 EXPORT_SYMBOL(ss_connect);
 
 /*
- * The original functions are inet_bind() and inet6_bind().
- * These two can be made a bit shorter should that become necessary.
+ * Bind for IPv4 protocol. The original functions are inet_bind().
+ *
+ * NOTE: This code assumes that both IPv4 and IPv6 are compiled in as
+ * part of the Linux kernel, and not as separate loadable kernel modules.
  */
+static int ss_inet_bind(struct sock *sk, struct sockaddr *uaddr, int addr_len)
+{
+	struct sockaddr_in *addr = (struct sockaddr_in *)uaddr;
+	struct inet_sock *inet = inet_sk(sk);
+	struct net *net = sock_net(sk);
+	unsigned short snum;
+	int chk_addr_ret;
+	int err;
+
+	err = -EINVAL;
+	if (addr_len < sizeof(struct sockaddr_in))
+		goto out;
+
+	chk_addr_ret = inet_addr_type(net, addr->sin_addr.s_addr);
+
+	/* Not specified by any standard per-se, however it breaks too
+	 * many applications when removed.  It is unfortunate since
+	 * allowing applications to make a non-local bind solves
+	 * several problems with systems using dynamic addressing.
+	 * (ie. your servers still start up even if your ISDN link
+	 *  is temporarily down)
+	 */
+	err = -EADDRNOTAVAIL;
+	if (!sysctl_ip_nonlocal_bind &&
+	    !(inet->freebind || inet->transparent) &&
+	    addr->sin_addr.s_addr != htonl(INADDR_ANY) &&
+	    chk_addr_ret != RTN_LOCAL &&
+	    chk_addr_ret != RTN_MULTICAST &&
+	    chk_addr_ret != RTN_BROADCAST)
+		goto out;
```

Comment:
As a matter of fact, we do set `inet_sk(sk)->freebind = 1;` in `sock_clnt.c`, so the condition above is never true. Do we need to keep this code block?